### PR TITLE
Init CI, nix shells and an MLIR derivation

### PR DESCRIPTION
**Pull request by doug-q**
While reviewing, please do review the repository settings. I have intended to copy them from hugr, except that I am allowing PRs to rebase.

See the develop branch for where this is going.

@croyzor may I ask you to review?

### CODE REVIEW

File: nix/mlir/updated-llvm-lit-cfg-add-libs-to-dylib-path.patch [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+@@ -41,6 +41,26 @@ llvm_config.with_environment("PATH", config.llvm_tools_dir, append_path=True)
+ # Propagate some variables from the host environment.
+ llvm_config.with_system_environment(["HOME", "INCLUDE", "LIB", "TMP", "TEMP"])
+ 
++# Add the LLVM dynamic libs to the platform-specific loader search path env var:
++#
++# TODO: this is copied from `clang`'s `lit.cfg.py`; should unify..
++def find_shlibpath_var():
++    if platform.system() in ['Linux', 'FreeBSD', 'NetBSD', 'OpenBSD', 'SunOS']:
++        yield 'LD_LIBRARY_PATH'
```
Comment (croyzor) [inline nix/mlir/updated-llvm-lit-cfg-add-libs-to-dylib-path.patch:26]: Why is this written as a generator? We could move the `else` branch below into a catch all case and avoid the for loop

File: nix/mlir/updated-llvm-lit-cfg-add-libs-to-dylib-path.patch [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+ # Set up OCAMLPATH to include newly built OCaml libraries.
+ top_ocaml_lib = os.path.join(config.llvm_lib_dir, "ocaml")
+@@ -59,6 +79,26 @@ llvm_config.with_environment("OCAMLRUNPARAM", "b")
+ # Provide the path to asan runtime lib 'libclang_rt.asan_osx_dynamic.dylib' if
+ # available. This is darwin specific since it's currently only needed on darwin.
+ 
++# Add the LLVM dynamic libs to the platform-specific loader search path env var:
++#
++# TODO: this is copied from `clang`'s `lit.cfg.py`; should unify..
++def find_shlibpath_var():
```
Comment (croyzor) [inline nix/mlir/updated-llvm-lit-cfg-add-libs-to-dylib-path.patch:51]: The same function is being added twice to the same file?

**Comment by croyzor** [comment]
As for the repository settings, I don't have permission to view them or add myself as a reviewer, so something is missing!

File: nix/mlir/default.nix [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        let go = x: if x.name == python3.name then python3 else x;
+        in map go old.nativeBuildInputs;
+
+      # This overrides the python that is provided by the base derivation
+      passthru = old.passthru or { } // { python = python3; };
+
+      # These are patches from nixpkgs, adapted to the newer llvm we are
+      # building against.
+      # If it becomes annoying to maintain these for multiple llvm versions we
+      # can take these as parameters.
```
Comment (ss2165) [inline nix/mlir/default.nix:68]: Could you add the sources of each patch and a comment saying what each one does

**Review by ss2165 — COMMENTED** [review]
👍

**Comment by ss2165** [comment]
Repository settings seem fine

File: nix/mlir/updated-llvm-lit-cfg-add-libs-to-dylib-path.patch (hunk 1 above)
Comment (doug-q) [reply nix/mlir/updated-llvm-lit-cfg-add-libs-to-dylib-path.patch:26]: I don't disagree but this is how it came from upstream (nixpkgs/pkgs/development/compilers/llvm/git/llvm/llvm-lit-cfg-add-libs-to-dylib-path.patch). I think it's best to leave it as is.

File: nix/mlir/updated-llvm-lit-cfg-add-libs-to-dylib-path.patch (hunk 2 above)
Comment (doug-q) [reply nix/mlir/updated-llvm-lit-cfg-add-libs-to-dylib-path.patch:51]: Agree, will remove the second.

File: nix/mlir/default.nix (hunk 3 above)
Comment (doug-q) [reply nix/mlir/default.nix:68]: Will do, I will copy the comments from nixpkgs.

File: .gitignore [hunk 4]
```diff
@@ -0,0 +1,9 @@
+/[_]build*/
```
[the remark this answers is not in the document]
Comment (doug-q) [reply .gitignore:1]: I don't think it's redundant: I like to use "/_build-blah" for build directories and this is the only pattern that matches that.

[the remark this answers is not in the document]
Comment (croyzor) [reply .gitignore:1]: This is extreme nitpicking, but what's wrong with `/_build*/`?

[the remark this answers is not in the document]
Comment (doug-q) [reply .gitignore:1]: Awkward, I somehow had thought the square brackets meant optional here. Thanks :)

**Comment by doug-q** [comment]
@croyzor Can you look again?

**Review by croyzor — APPROVED** [review]

**Comment by croyzor** [comment]
> At least 1 approving review is required by reviewers with write access.

As discussed, we should default to giving write access to the `tket` team for these repos. Once that's done, you should be able to merge! 🙂

**Pull request merged**